mmcv/ops/csrc/pytorch/cuda/filtered_lrelu.cu

```diff
@@ -0,0 +1,2045 @@
+  MODE_SUFD = 2,  // Separable upsampling, full downsampling.
+  MODE_FUFD = 3,  // Full upsampling, full downsampling.
+};
+
+template <class T>
+struct InternalType;
+template <>
+struct InternalType<double> {
+  typedef double scalar_t;
+  typedef double2 vec2_t;
+  typedef double4 vec4_t;
+  __device__ __forceinline__ static vec2_t zero_vec2(void) {
+    return make_double2(0, 0);
+  }
+  __device__ __forceinline__ static vec4_t zero_vec4(void) {
+    return make_double4(0, 0, 0, 0);
+  }
+  __device__ __forceinline__ static double clamp(double x, double c) {
+    return fmin(fmax(x, -c), c);
+  }
+};
+template <>
+struct InternalType<float> {
+  typedef float scalar_t;
+  typedef float2 vec2_t;
+  typedef float4 vec4_t;
+  __device__ __forceinline__ static vec2_t zero_vec2(void) {
+    return make_float2(0, 0);
+  }
+  __device__ __forceinline__ static vec4_t zero_vec4(void) {
+    return make_float4(0, 0, 0, 0);
+  }
+  __device__ __forceinline__ static float clamp(float x, float c) {
+    return fminf(fmaxf(x, -c), c);
+  }
+};
+template <>
+struct InternalType<c10::Half> {
+  typedef float scalar_t;
+  typedef float2 vec2_t;
+  typedef float4 vec4_t;
+  __device__ __forceinline__ static vec2_t zero_vec2(void) {
+    return make_float2(0, 0);
+  }
+  __device__ __forceinline__ static vec4_t zero_vec4(void) {
+    return make_float4(0, 0, 0, 0);
+  }
+  __device__ __forceinline__ static float clamp(float x, float c) {
+    return fminf(fmaxf(x, -c), c);
+  }
+};
+
+#define MIN(A, B) ((A) < (B) ? (A) : (B))
+#define MAX(A, B) ((A) > (B) ? (A) : (B))
+#define CEIL_DIV(A, B) \
+  (((B) == 1)
+       ? (A)                   \
+       : ((B) == 2) ? ((int)((A) + 1) >> 1) \
+                    : ((B) == 4) ? ((int)((A) + 3) >> 2) \
+                                 : (((A) + ((A) > 0 ? (B)-1 : 0)) / (B)))
+
+// This works only up to blocks of size 256 x 256 and for all N that are powers
+// of two.
+template <int N>
+__device__ __forceinline__ void fast_div_mod(int& x, int& y, unsigned int i) {
+         if ((N & (N - 1)) && N <= 256)
+           y = (i * ((1 << 24) / N + 1)) >> 24;  // Assumes N <= 256, i < N*256.
+         else
+           y = i / N;
+
+         x = i - y * N;
+       }
+
+       // Type cast stride before reading it.
+       template <class T>
+       __device__ __forceinline__ T get_stride(const int64_t& x) {
+         return *reinterpret_cast<const T*>(&x);
+       }
+
+       //------------------------------------------------------------------------
+       // Filters, setup kernel, copying function.
+
+#define MAX_FILTER_SIZE 32
+
+       // Combined up/down filter buffers so that transfer can be done with one
+       // copy.
+       __device__ float g_fbuf[2 * MAX_FILTER_SIZE *
+                               MAX_FILTER_SIZE];  // Filters in global memory,
+                                                  // written by setup kernel.
+       __device__ __constant__ float
```

Comment:
Constant memory size is small, only 64kb. I am not sure if it is a good idea to use constant memory for this op.

Reply:
This part is directly from https://github.com/NVlabs/stylegan3/blob/main/torch_utils/ops/filtered_lrelu.cu#L77. Maybe keeping it as origin is fine.